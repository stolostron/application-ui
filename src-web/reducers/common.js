/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2017, 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
/*
Since these selectors are common (i.e used across different components and different parts of the store)
we have to create unique selectors each invocation and the selectors need additional metata data sent
though props that indicate which part of the store it should select from.

//See https://github.com/reactjs/reselect#sharing-selectors-with-props-across-multiple-components
//See https://github.com/reactjs/reselect#q-can-i-share-a-selector-across-multiple-components

The selector pattern is an abstraction that standardizes an applicationâ€™s store querying logic.
It is simple: for any part of the store that an application needs access to, define a function that
when given the full store, returns the desired part (or derivation) of the store.
*/

import { createSelector } from 'reselect'
import lodash from 'lodash'
import { normalize } from 'normalizr'
import ReactDOMServer from 'react-dom/server'
import { createResourcesSchema } from '../../lib/client/resource-schema'
import { transform } from '../../lib/client/resource-helper'
import * as Actions from '../actions'
import getResourceDefinitions, * as ResourceDefinitions from '../definitions'
import { RESOURCE_TYPES } from '../../lib/shared/constants'
import msgs from '../../nls/platform.properties'

export const getItems = (state, props) =>
  getFromState(state, props.storeRoot, 'items')
export const getItemsPerPage = (state, props) =>
  getFromState(state, props.storeRoot, 'itemsPerPage')
export const getPage = (state, props) =>
  getFromState(state, props.storeRoot, 'page')
export const getSearch = (state, props) =>
  getFromState(state, props.storeRoot, 'search')
export const getSortColumn = (state, props) =>
  getFromState(state, props.storeRoot, 'sortColumn')
export const getSortDirection = (state, props) =>
  getFromState(state, props.storeRoot, 'sortDirection')

function getFromState(state, root, attribute) {
  const storeRoot = state[root]
  if (storeRoot) {
    return storeRoot[attribute]
  } else {
    //eslint-disable-next-line no-console
    console.error(`store root '${root}' does not exist`)
    return undefined
  }
}

let globalContext = { locale: 'en-US' }
try {
  globalContext = JSON.parse(document.getElementById('context').textContent)
} catch (e) {
  globalContext = { locale: 'en-US' }
}

export const INITIAL_STATE = {
  items: [],
  itemsPerPage: Actions.PAGE_SIZES.DEFAULT,
  page: 1,
  search: '',
  sortColumn: undefined,
  sortDirection: Actions.SORT_DIRECTION_ASCENDING,
  status: Actions.REQUEST_STATUS.INCEPTION,
  putStatus: undefined,
  putErrorMsg: '',
  postStatus: undefined,
  postStatusCode: undefined,
  postErrorMsg: '',
  pendingActions: [],
  clientSideFilters: undefined,
  forceReload: false,
  mutateStatus: undefined,
  mutateErrorMsg: undefined,
  deleteMsg: undefined,
  deleteStatus: undefined
}

/**
 * Search in a table cell.  Uses the transformed data for the cell.
 *
 * @param {*} item - Table row data
 * @param {*} tableKey - Table column key
 * @param {*} context - React context
 * @param {*} searchText - String to match
 */
function searchTableCell(item, tableKey, context, searchText) {
  const renderedElement = transform(item, tableKey, context.locale, true)
  if (typeof renderedElement === 'string') {
    return (
      renderedElement.toLowerCase().indexOf(searchText.toLowerCase()) !== -1
    )
  } else {
    return (
      ReactDOMServer.renderToString(
        transform(item, tableKey, context.locale, true)
      )
        .toString()
        .toLowerCase()
        .indexOf(searchText.toLowerCase()) !== -1
    )
  }
}

function searchTableCellHelper(search, tableKeys, item, context) {
  const searchKey = search.substring(0, search.indexOf('='))
  const searchField = search.substring(search.indexOf('=') + 1)

  if (searchKey === 'textsearch') {
    return tableKeys.find(tableKeyTmp =>
      searchTableCell(
        item,
        tableKeyTmp,
        context,
        searchField.replace(/[{}]/g, '')
      )
    )
  }
  const tableKey = tableKeys.find(
    tableKeyTmp =>
      msgs.get(tableKeyTmp.msgKey, context.locale).toLowerCase() ===
      searchKey.toLowerCase()
  )
  if (!lodash.isEmpty(searchField)) {
    if (!searchField.includes('{')) {
      if (tableKey) {
        return searchTableCell(item, tableKey, context, searchField)
      }
    } else {
      let found = false
      const searchKeys = searchField.replace(/[{}]/g, '').split(',')
      if (searchKeys && tableKey) {
        searchKeys.forEach(searchKeyTmp => {
          if (searchTableCell(item, tableKey, context, searchKeyTmp)) {
            found = true
          }
        })
      }
      return found
    }
  }
  // return all results when user types cluster=
  if (searchField === '') {
    return true
  }

  // by default, search all fields
  return tableKeys.find(tableKeyTmp =>
    searchTableCell(item, tableKeyTmp, context, search)
  )
}

const makeGetFilteredItemsSelector = resourceType => {
  return createSelector([getItems, getSearch], (items, search) =>
    items.filter(item => {
      if (lodash.isEmpty(search)) {
        return true
      }

      const tableKeys = ResourceDefinitions.getTableKeys(resourceType)

      if (search.includes('},')) {
        // special case like status={healthy}, labels={cloud=IBM}
        let found = false
        const searchFields = search
          .replace('},', '}},')
          .toLowerCase()
          .split('},')
        searchFields.forEach(searchField => {
          if (
            searchTableCellHelper(searchField, tableKeys, item, globalContext)
          ) {
            found = true
          }
        })
        return found
      } else {
        return searchTableCellHelper(search, tableKeys, item, globalContext)
      }
    })
  )
}

const makeGetTransformedItemsSelector = resourceType => {
  return createSelector([makeGetFilteredItemsSelector(resourceType)], items => {
    const resourceData = getResourceDefinitions(resourceType)
    return items.map(item => {
      const customData = {}

      resourceData.tableKeys.forEach(key => {
        if (
          key.transformFunction &&
          typeof key.transformFunction === 'function'
        ) {
          customData[
            key.resourceKey.replace('custom.', '')
          ] = key.transformFunction(item, globalContext.locale)
        }
      })
      item.custom = customData
      return item
    })
  })
}

//Could we do better? - we have one selector thus one cache for sorting.
//Thus if the sort direction change is toggled back we re-calculate.
const makeGetSortedItemsSelector = resourceType => {
  return createSelector(
    [
      makeGetTransformedItemsSelector(resourceType),
      getSortColumn,
      getSortDirection
    ],
    (items, sortColumn, sortDirection) => {
      const initialSortField = sortColumn
        ? sortColumn
        : ResourceDefinitions.getDefaultSortField(resourceType)
      const sortField =
        initialSortField === 'custom.age'
          ? 'metadata.creationTimestamp'
          : initialSortField // sort by the actual date, not formatted value
      const direction =
        sortDirection === Actions.SORT_DIRECTION_ASCENDING
          ? Actions.SORT_DIRECTION_DESCENDING
          : Actions.SORT_DIRECTION_ASCENDING
      const sortDir =
        sortField === 'metadata.creationTimestamp' ? direction : sortDirection // date fields should initially sort from latest to oldest
      return lodash.orderBy(items, item => lodash.get(item, sortField), [
        sortDir
      ])
    }
  )
}

const makeGetPagedItemsSelector = resourceType => {
  return createSelector(
    [makeGetSortedItemsSelector(resourceType), getPage, getItemsPerPage],
    (items, page, itemsPerPage) => {
      const offset = (page - 1) * itemsPerPage
      let lastIndex = offset + itemsPerPage
      lastIndex = lastIndex <= items.length ? lastIndex : items.length
      return {
        items: items.slice(offset, lastIndex),
        totalResults: items.length,
        totalPages: Math.ceil(items.length / itemsPerPage)
      }
    }
  )
}

export const makeGetVisibleTableItemsSelector = resourceType => {
  const pk = ResourceDefinitions.getPrimaryKey(resourceType)
  const sk = ResourceDefinitions.getSecondaryKey(resourceType)
  return createSelector([makeGetPagedItemsSelector(resourceType)], result => {
    const normalizedItems = normalize(result.items, [
      createResourcesSchema(pk, sk)
    ]).entities.items
    return Object.assign(result, {
      normalizedItems: normalizedItems,
      items: result.items.map(
        item =>
          sk
            ? `${lodash.get(item, pk)}-${lodash.get(item, sk)}`
            : `${lodash.get(item, pk)}`
      ) // to support multi cluster, use ${name}-${cluster} as unique id
    })
  })
}

export const secondaryHeader = (
  state = { title: '', tabs: [], breadcrumbItems: [], links: [] },
  action
) => {
  switch (action.type) {
  case Actions.SECONDARY_HEADER_UPDATE:
    return Object.assign({}, state, {
      title: action.title,
      tabs: action.tabs,
      breadcrumbItems: action.breadcrumbItems,
      links: action.links,
      actions: action.actions,
      tooltip: action.tooltip
    })
  default:
    return state
  }
}

const getItemProps = (state, props) => props

export const getSingleResourceItem = createSelector(
  [getItems, getItemProps],
  (items, props) => items && items.length > 0 && props.predicate(items, props)
)

export const resourceItemByName = (items, props) => {
  const key = ResourceDefinitions.getURIKey(props.resourceType)
  return lodash.find(items, item => lodash.get(item, key) === props.name)
}

export const resourceItemByNameAndNamespace = (items, props) => {
  const key = ResourceDefinitions.getURIKey(props.resourceType)
  return lodash.find(
    items,
    item =>
      lodash.get(item, key) === props.name &&
      lodash.get(item, 'metadata.namespace') === props.namespace
  )
}

export const resourceReducerFunction = (state = INITIAL_STATE, action) => {
  let items, index
  switch (action.type) {
  case Actions.RESOURCE_REQUEST:
    return Object.assign({}, state, {
      status: Actions.REQUEST_STATUS.IN_PROGRESS
    })
  case Actions.RESOURCE_RECEIVE_SUCCESS: {
    const pageSize =
        action.items.length > state.itemsPerPage * (state.page - 1)
          ? state.page
          : state.page - 1
    return Object.assign({}, state, {
      status: Actions.REQUEST_STATUS.DONE,
      items: action.items,
      page: action.items.length === 0 ? 1 : pageSize,
      resourceVersion: action.resourceVersion
    })
  }
  case Actions.RESOURCE_RECEIVE_FAILURE:
    return Object.assign({}, state, {
      status: Actions.REQUEST_STATUS.ERROR,
      err: action.err
    })
  case Actions.POST_REQUEST:
    return Object.assign({}, state, {
      postStatus: Actions.REQUEST_STATUS.IN_PROGRESS
    })
  case Actions.POST_RECEIVE_SUCCESS:
    items = state.items.slice(0)
    if (action.item.length > 0) {
      // if returned as an array due to making async calls, push action.item elements to the items array
      action.item.forEach(el => {
        items.push(el)
      })
    } else {
      items.push(action.item)
    }
    return Object.assign({}, state, {
      items: items,
      postStatus: Actions.REQUEST_STATUS.DONE
    })
  case Actions.POST_RECEIVE_FAILURE:
    return Object.assign({}, state, {
      postStatus: Actions.REQUEST_STATUS.ERROR,
      postStatusCode:
          action.err.error &&
          action.err.error.response &&
          action.err.error.response.status,
      postErrorMsg: action.err.error && action.err.error.message
    })
  case Actions.PUT_REQUEST:
    return Object.assign({}, state, {
      putStatus: Actions.REQUEST_STATUS.IN_PROGRESS
    })
  case Actions.PUT_RECEIVE_SUCCESS:
    return Object.assign({}, state, {
      putStatus: Actions.REQUEST_STATUS.DONE
    })
  case Actions.PUT_RECEIVE_FAILURE:
    return Object.assign({}, state, {
      putStatus: Actions.REQUEST_STATUS.ERROR,
      putErrorMsg: action.err.error
        ? action.err.error.message
        : action.err.message
    })
  case Actions.CLEAR_REQUEST_STATUS:
    return Object.assign({}, state, {
      mutateStatus: undefined,
      mutateErrorMsg: undefined,
      postStatus: undefined,
      postStatusCode: undefined,
      postErrorMsg: undefined,
      putStatus: undefined,
      putErrorMsg: undefined,
      deleteMsg: undefined,
      deleteStatus: undefined
    })
  case Actions.TABLE_SEARCH:
    return Object.assign({}, state, {
      search: action.search,
      page: 1
    })
  case Actions.TABLE_SORT:
    return Object.assign({}, state, {
      sortDirection: action.sortDirection,
      sortColumn: action.sortColumn
    })
  case Actions.TABLE_PAGE_CHANGE:
    return Object.assign({}, state, {
      page: action.page,
      itemsPerPage: action.pageSize
    })
  case Actions.RESOURCE_ADD:
  case Actions.RESOURCE_MODIFY: {
    const resourceTypeObj = !lodash.isObject(action.resourceType)
      ? RESOURCE_TYPES[
        lodash.findKey(RESOURCE_TYPES, { name: action.resourceType })
      ]
      : action.resourceType
    const primaryKey = ResourceDefinitions.getPrimaryKey(resourceTypeObj)
    items = state.items.slice(0)
    index = lodash.findIndex(
      items,
      o => lodash.get(o, primaryKey) === lodash.get(action.item, primaryKey)
    )
    index > -1
      ? items.splice(index, 1, action.item)
      : items.push(action.item)
    return Object.assign({}, state, {
      items: items
    })
  }
  case Actions.RESOURCE_MUTATE:
    return Object.assign({}, state, {
      mutateStatus: Actions.REQUEST_STATUS.IN_PROGRESS,
      mutateErrorMsg: null,
      pendingActions: [
        ...state.pendingActions,
        { name: action.resourceName, action: Actions.RESOURCE_MUTATE }
      ]
    })
  case Actions.RESOURCE_MUTATE_FAILURE:
    return Object.assign({}, state, {
      mutateStatus: Actions.REQUEST_STATUS.ERROR,
      mutateErrorMsg:
          action.err.message ||
          (action.err.error &&
            (action.err.error.message ||
              (action.err.error.data && action.err.error.data.Message))),
      pendingActions: state.pendingActions.filter(
        r => r && r.name !== action.resourceName
      )
    })
  case Actions.RESOURCE_MUTATE_SUCCESS:
    return Object.assign({}, state, {
      mutateStatus: Actions.REQUEST_STATUS.DONE,
      pendingActions: state.pendingActions.filter(
        r => r && r.name !== action.resourceName
      )
    })
  case Actions.RESOURCE_DELETE:
    items = [...state.items]
    index = lodash.findIndex(
      items,
      o =>
        lodash.get(o, 'metadata.uid') ===
          lodash.get(action, 'item.metadata.uid')
    )
    if (index > -1) {
      items.splice(index, 1)
      return Object.assign({}, state, {
        deleteStatus: Actions.REQUEST_STATUS.IN_PROGRESS,
        deleteMsg: null,
        items: items
      })
    }
    return state
  case Actions.DEL_RECEIVE_SUCCESS: {
    items = [...state.items]

    const { item } = action
    if (item && item.name && item.namespace) {
      index = lodash.findIndex(
        items,
        o =>
          lodash.get(o, 'name') === lodash.get(item, 'name') &&
            lodash.get(o, 'namespace') === lodash.get(item, 'namespace')
      )
      if (index > -1) {
        items.splice(index, 1)
      }
    }

    return Object.assign({}, state, {
      items: items,
      deleteStatus: Actions.REQUEST_STATUS.DONE,
      deleteMsg: action.item.name
    })
  }
  case Actions.RESOURCE_FORCE_RELOAD_FINISHED:
    return Object.assign({}, state, {
      forceReload: false
    })
  case Actions.DEL_RECEIVE_SUCCESS_FINISHED:
    return Object.assign({}, state, {
      deleteStatus: undefined,
      deleteMsg: null
    })
  case Actions.RESOURCE_MUTATE_FINISHED:
    return Object.assign({}, state, {
      mutateStatus: undefined
    })
  case Actions.RESOURCE_FORCE_RELOAD:
    return Object.assign({}, state, {
      forceReload: true
    })
  default:
    return state
  }
}

/**
 * A common higher order reducer for resources
 * Follows the Redux pattern described here: http://redux.js.org/docs/recipes/reducers/ReusingReducerLogic.html#customizing-behavior-with-higher-order-reducers
 * **/
export const createResourceReducer = (reducerFunction, reducerPredicate) => {
  return (state, action) => {
    const isInitializationCall = state === undefined
    const shouldRunWrappedReducer =
      reducerPredicate(action) || isInitializationCall
    return shouldRunWrappedReducer ? reducerFunction(state, action) : state
  }
}
