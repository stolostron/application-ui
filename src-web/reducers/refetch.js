/*******************************************************************************
 * Copyright (c) 2020 Red Hat, Inc.
 *******************************************************************************/

import { REFETCH_INTERVAL_UPDATE } from '../actions'

export const refetch = (state, action) => {
  switch (action.type) {
  case REFETCH_INTERVAL_UPDATE:
    return action.data
  default:
    // get the poll interval from RefreshTimeSelect component
    return Object.assign({}, state, {})
  }
}
