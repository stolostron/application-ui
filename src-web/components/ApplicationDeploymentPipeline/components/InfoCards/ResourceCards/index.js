/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/

import React from 'react'
import { connect } from 'react-redux'
import resources from '../../../../../../lib/shared/resources'
import { RESOURCE_TYPES } from '../../../../../../lib/shared/constants'
import { fetchResources } from '../../../../../actions/common'
import msgs from '../../../../../../nls/platform.properties'
import {
  getNumPlacementRules,
  getSubscriptionDataOnHub,
  getSubscriptionDataOnManagedClustersSingle,
  getSubscriptionDataOnManagedClustersRoot
} from '../utils'
import {
  getSearchLinkForOneApplication,
  getSearchLinkForAllSubscriptions,
  getSearchLinkForAllClusters,
  getSearchLinkForAllChannels,
  getSearchLinkForAllPlacementRules
} from '../../../../common/ResourceOverview/utils'

/* eslint-disable react/prop-types */

resources(() => {
  require('../style.scss')
})

const mapDispatchToProps = dispatch => {
  return {
    fetchChannels: () => dispatch(fetchResources(RESOURCE_TYPES.HCM_CHANNELS))
  }
}

const mapStateToProps = state => {
  const { HCMSubscriptionList, QueryApplicationList } = state
  return {
    HCMSubscriptionList,
    QueryApplicationList
  }
}

const getResourceCardsData = (
  HCMSubscriptionList,
  QueryApplicationList,
  globalAppData,
  isSingleApplicationView,
  applicationName,
  applicationNamespace,
  targetLinkForSubscriptions,
  targetLinkForClusters,
  targetLinkForChannels,
  targetLinkForPlacementRules,
  locale
) => {
  const subscriptionDataOnHub = getSubscriptionDataOnHub(
    QueryApplicationList,
    isSingleApplicationView,
    applicationName,
    applicationNamespace
  )

  let subscriptionDataOnManagedClusters
  if (isSingleApplicationView) {
    subscriptionDataOnManagedClusters = getSubscriptionDataOnManagedClustersSingle(
      QueryApplicationList,
      applicationName,
      applicationNamespace
    )
  } else {
    subscriptionDataOnManagedClusters = getSubscriptionDataOnManagedClustersRoot(
      globalAppData
    )
  }

  const placementRules = getNumPlacementRules(
    HCMSubscriptionList,
    isSingleApplicationView,
    applicationNamespace
  )

  const result = [
    {
      msgKey:
        subscriptionDataOnHub.total == 1
          ? msgs.get('dashboard.card.deployment.subscription', locale)
          : msgs.get('dashboard.card.deployment.subscriptions', locale),
      count: subscriptionDataOnHub.total,
      targetLink:
        subscriptionDataOnHub.total == 0 ? '' : targetLinkForSubscriptions,
      textKey: msgs.get('dashboard.card.deployment.subscriptions.text', locale),
      subtextKeyFirst:
        subscriptionDataOnHub.total > 0
          ? subscriptionDataOnHub.failed
            .toString()
            .concat(
              ' ',
              msgs.get('dashboard.card.deployment.failed.lowercase', locale)
            )
          : '',
      subtextKeySecond:
        subscriptionDataOnHub.noStatus > 0
          ? subscriptionDataOnHub.noStatus
            .toString()
            .concat(
              ' ',
              msgs.get('dashboard.card.deployment.noStatus', locale)
            )
          : ''
    },
    {
      msgKey:
        subscriptionDataOnManagedClusters.clusters == 1
          ? msgs.get('dashboard.card.deployment.managedCluster', locale)
          : msgs.get('dashboard.card.deployment.managedClusters', locale),
      count: subscriptionDataOnManagedClusters.clusters,
      targetLink:
        subscriptionDataOnManagedClusters.clusters == 0
          ? ''
          : targetLinkForClusters,
      textKey: subscriptionDataOnManagedClusters.total
        .toString()
        .concat(
          ' ',
          subscriptionDataOnManagedClusters.total == 1
            ? msgs.get('dashboard.card.deployment.totalSubscription', locale)
            : msgs.get('dashboard.card.deployment.totalSubscriptions', locale)
        ),
      subtextKeyFirst:
        subscriptionDataOnManagedClusters.clusters > 0
          ? subscriptionDataOnManagedClusters.failed
            .toString()
            .concat(
              ' ',
              msgs.get('dashboard.card.deployment.failed.lowercase', locale)
            )
          : '',
      subtextKeySecond:
        subscriptionDataOnManagedClusters.noStatus > 0
          ? subscriptionDataOnManagedClusters.noStatus
            .toString()
            .concat(
              ' ',
              msgs.get('dashboard.card.deployment.noStatus', locale)
            )
          : ''
    },
    {
      msgKey:
        subscriptionDataOnHub.channels == 1
          ? msgs.get('dashboard.card.deployment.channel', locale)
          : msgs.get('dashboard.card.deployment.channels', locale),
      count: subscriptionDataOnHub.channels,
      targetLink:
        subscriptionDataOnHub.channels == 0 ? '' : targetLinkForChannels,
      textKey: isSingleApplicationView
        ? msgs.get('dashboard.card.deployment.used', locale)
        : msgs.get('dashboard.card.deployment.total', locale)
    },
    {
      msgKey:
        placementRules == 1
          ? msgs.get('dashboard.card.deployment.placementRule', locale)
          : msgs.get('dashboard.card.deployment.placementRules', locale),
      count: placementRules,
      targetLink: placementRules == 0 ? '' : targetLinkForPlacementRules,
      textKey: isSingleApplicationView
        ? msgs.get('dashboard.card.deployment.used', locale)
        : msgs.get('dashboard.card.deployment.total', locale)
    }
  ]

  return result
}

class ResourceCards extends React.Component {
  componentWillMount() {}
  componentDidMount() {}

  componentWillUnmount() {}

  render() {
    const {
      HCMSubscriptionList,
      QueryApplicationList,
      isSingleApplicationView,
      selectedAppName,
      selectedAppNS,
      globalAppData
    } = this.props
    const { locale } = this.context
    const singleAppStyle = isSingleApplicationView ? ' single-app' : ''
    const applicationName = selectedAppName
    const applicationNamespace = selectedAppNS

    const targetLinkForSubscriptions = isSingleApplicationView
      ? getSearchLinkForOneApplication({
        name: encodeURIComponent(applicationName),
        showRelated: 'subscription'
      })
      : getSearchLinkForAllSubscriptions()
    const targetLinkForClusters = isSingleApplicationView
      ? getSearchLinkForOneApplication({
        name: encodeURIComponent(applicationName),
        showRelated: 'cluster'
      })
      : getSearchLinkForAllClusters()
    const targetLinkForChannels = isSingleApplicationView
      ? getSearchLinkForOneApplication({
        name: encodeURIComponent(applicationName),
        showRelated: 'channel'
      })
      : getSearchLinkForAllChannels()
    const targetLinkForPlacementRules = isSingleApplicationView
      ? getSearchLinkForOneApplication({
        name: encodeURIComponent(applicationName),
        showRelated: 'placementrule'
      })
      : getSearchLinkForAllPlacementRules()

    const resourceCardsData = getResourceCardsData(
      HCMSubscriptionList,
      QueryApplicationList,
      globalAppData,
      isSingleApplicationView,
      applicationName,
      applicationNamespace,
      targetLinkForSubscriptions,
      targetLinkForClusters,
      targetLinkForChannels,
      targetLinkForPlacementRules,
      locale
    )

    const handleClick = (e, resource) => {
      if (resource.targetLink) {
        window.open(resource.targetLink, '_blank')
      }
    }
    const handleKeyPress = (e, resource) => {
      if (e.key === 'Enter') {
        handleClick(e, resource)
      }
    }

    return (
      <div className={'resource-cards-info' + singleAppStyle}>
        {Object.keys(resourceCardsData).map(key => {
          const card = resourceCardsData[key]
          return (
            <React.Fragment key={key}>
              <div
                key={card}
                className={
                  card.targetLink ? 'single-card clickable' : 'single-card'
                }
                role="button"
                tabIndex="0"
                onClick={e => handleClick(e, card)}
                onKeyPress={e => handleKeyPress(e, card)}
              >
                <div className="card-count">{card.count}</div>
                <div className="card-type">{card.msgKey}</div>
                <div className="card-text">{card.textKey}</div>
                {(card.subtextKeyFirst || card.subtextKeySecond) && (
                  <div className="row-divider" />
                )}
                {card.subtextKeyFirst && (
                  <div className="card-subtext">{card.subtextKeyFirst}</div>
                )}
                {card.subtextKeySecond && (
                  <div className="card-subtext">{card.subtextKeySecond}</div>
                )}
              </div>
              {key < Object.keys(resourceCardsData).length - 1 && (
                <div className="column-divider" />
              )}
            </React.Fragment>
          )
        })}
      </div>
    )
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(ResourceCards)
