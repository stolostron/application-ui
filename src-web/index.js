/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2017, 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
import React from 'react'
import { hydrate } from 'react-dom'
import { Provider } from 'react-redux'
import thunkMiddleware from 'redux-thunk'
import { createLogger } from 'redux-logger'
import { createStore, combineReducers, applyMiddleware, compose } from 'redux'
import { Router } from 'react-router-dom'
import { ApolloProvider } from 'react-apollo'
import App from './containers/App'
import ScrollToTop from './components/common/ScrollToTop'
import * as reducers from './reducers'
import config from '../lib/shared/config'
import apolloClient from '../lib/client/apollo-client'
import history from 'history'

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose

window.SHARED_HISTORY = window.SHARED_HISTORY
  ? window.SHARED_HISTORY
  : history.createBrowserHistory()

const loggerMiddleware = createLogger()
// Grab the state from a global variable injected into the server-generated HTML
const preloadedState = window.__PRELOADED_STATE__

const middleware = [thunkMiddleware] // lets us dispatch() functions
if (
  !window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ &&
  config['featureFlags:reduxLogger']
) {
  middleware.push(loggerMiddleware) // middleware that logs actions
}

// Allow the passed state to be garbage-collected
delete window.__PRELOADED_STATE__

// Create Redux store with initial state
const store = createStore(
  combineReducers(reducers),
  preloadedState,
  composeEnhancers(applyMiddleware(...middleware))
)

hydrate(
  <ApolloProvider client={apolloClient.getClient()}>
    <Provider store={store}>
      <Router history={window.SHARED_HISTORY}>
        <ScrollToTop>
          <App />
        </ScrollToTop>
      </Router>
    </Provider>
  </ApolloProvider>,
  document.getElementById('page')
)
