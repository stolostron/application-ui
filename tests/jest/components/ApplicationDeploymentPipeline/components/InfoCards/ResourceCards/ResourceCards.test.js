/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/

const React = require("../../../../../../../node_modules/react");

import ResourceCards from "../../../../../../../src-web/components/ApplicationDeploymentPipeline/components/InfoCards/ResourceCards";

import renderer from "react-test-renderer";
import * as reducers from "../../../../../../../src-web/reducers";

import { createStore, combineReducers, applyMiddleware, compose } from "redux";
import thunkMiddleware from "redux-thunk";
import { Provider } from "react-redux";

const preloadedState = window.__PRELOADED_STATE__;
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
const middleware = [thunkMiddleware];

const store = createStore(
  combineReducers(reducers),
  preloadedState,
  composeEnhancers(applyMiddleware(...middleware))
);

describe("ResourceCards", () => {
  it("ResourceCards renders correctly in root app view.", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <ResourceCards
            selectedAppName="app1"
            selectedAppNS="default"
            isSingleApplicationView={false}
            globalAppData={globalAppData}
          />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("ResourceCards renders correctly in single app view.", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <ResourceCards
            selectedAppName="app1"
            selectedAppNS="default"
            isSingleApplicationView={true}
            globalAppData={globalAppData}
          />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

const globalAppData = {
  items: {
    channelsCount: 1,
    clusterCount: 2,
    hubSubscriptionCount: 3,
    remoteSubscriptionStatusCount: {
      Subscribed: 3,
      Failed: 2,
      null: 1
    }
  }
};
