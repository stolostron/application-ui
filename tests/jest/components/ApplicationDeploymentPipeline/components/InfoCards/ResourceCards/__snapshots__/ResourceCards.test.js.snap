// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceCards ResourceCards renders correctly in root app view. 1`] = `
<div
  className="resource-cards-info"
>
  <div
    className="single-card"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <div
      className="card-count"
    >
      0
    </div>
    <div
      className="card-type"
    >
      Subscriptions
    </div>
    <div
      className="card-text"
    >
      On hub cluster
    </div>
    
    
    
  </div>
  <div
    className="column-divider"
  />
  <div
    className="single-card clickable"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <div
      className="card-count"
    >
      2
    </div>
    <div
      className="card-type"
    >
      Managed clusters
    </div>
    <div
      className="card-text"
    >
      6 Total subscriptions
    </div>
    <div
      className="row-divider"
    />
    <div
      className="card-subtext"
    >
      2 failed
    </div>
    <div
      className="card-subtext"
    >
      1 no status
    </div>
  </div>
  <div
    className="column-divider"
  />
  <div
    className="single-card"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <div
      className="card-count"
    >
      0
    </div>
    <div
      className="card-type"
    >
      Channels
    </div>
    <div
      className="card-text"
    >
      Total
    </div>
  </div>
  <div
    className="column-divider"
  />
  <div
    className="single-card"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <div
      className="card-count"
    >
      0
    </div>
    <div
      className="card-type"
    >
      Placement rules
    </div>
    <div
      className="card-text"
    >
      Total
    </div>
  </div>
</div>
`;

exports[`ResourceCards ResourceCards renders correctly in single app view. 1`] = `
<div
  className="resource-cards-info single-app"
>
  <div
    className="single-card"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <div
      className="card-count"
    >
      0
    </div>
    <div
      className="card-type"
    >
      Subscriptions
    </div>
    <div
      className="card-text"
    >
      On hub cluster
    </div>
    
    
    
  </div>
  <div
    className="column-divider"
  />
  <div
    className="single-card"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <div
      className="card-count"
    >
      0
    </div>
    <div
      className="card-type"
    >
      Managed clusters
    </div>
    <div
      className="card-text"
    >
      0 Total subscriptions
    </div>
    
    
    
  </div>
  <div
    className="column-divider"
  />
  <div
    className="single-card"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <div
      className="card-count"
    >
      0
    </div>
    <div
      className="card-type"
    >
      Channels
    </div>
    <div
      className="card-text"
    >
      In use
    </div>
  </div>
  <div
    className="column-divider"
  />
  <div
    className="single-card"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <div
      className="card-count"
    >
      0
    </div>
    <div
      className="card-type"
    >
      Placement rules
    </div>
    <div
      className="card-text"
    >
      In use
    </div>
  </div>
</div>
`;
